src/measure_theory/conditional_expectation.lean

```diff
@@ -262,4 +273,87 @@ end complete_subspace
 
 end Lp_meas
 
+/-! ## Conditional expectation in L2
+
+We define a conditional expectation in `L2`: it is the orthogonal projection on the subspace
+`Lp_meas`. -/
+
+section condexp_L2
+
+local attribute [instance] fact_one_le_two_ennreal
+
+variables [complete_space E] [borel_space 𝕜] {m m0 : measurable_space α} {μ : measure α}
+  {s t : set α}
+
+local notation `⟪`x`, `y`⟫` := @inner 𝕜 E _ x y
+local notation `⟪`x`, `y`⟫₂` := @inner 𝕜 (α →₂[μ] E) _ x y
+
+variables (𝕜)
+/-- Conditional expectation of a function in L2 with respect to a sigma-algebra -/
+def condexp_L2 (hm : m ≤ m0) : (α →₂[μ] E) →L[𝕜] (Lp_meas E 𝕜 m 2 μ) :=
+@orthogonal_projection 𝕜 (α →₂[μ] E) _ _ (Lp_meas E 𝕜 m 2 μ)
+  (by { haveI : fact (m ≤ m0) := ⟨hm⟩, exact infer_instance, })
+variables {𝕜}
+
+lemma integrable_on_condexp_L2_of_measure_ne_top (hm : m ≤ m0) (hμs : μ s ≠ ∞) (f : α →₂[μ] E) :
+  integrable_on (condexp_L2 𝕜 hm f) s μ :=
+integrable_on_Lp_of_measure_ne_top ((condexp_L2 𝕜 hm f) : α →₂[μ] E)
+  fact_one_le_two_ennreal.elim hμs
+
+lemma integrable_condexp_L2_of_finite_measure (hm : m ≤ m0) [finite_measure μ] {f : α →₂[μ] E} :
+  integrable (condexp_L2 𝕜 hm f) μ :=
+integrable_on_univ.mp $ integrable_on_condexp_L2_of_measure_ne_top hm (measure_ne_top _ _) f
+
+lemma norm_condexp_L2_le_one (hm : m ≤ m0) : ∥@condexp_L2 α E 𝕜 _ _ _ _ _ _ _ _ _ _ μ hm∥ ≤ 1 :=
```

Comment:
I'm guessing you never worked out why you needed all these explicit arguments? I'm running into much worse versions of this problem in #8218...

Reply:
`E` is the argument that cannot be inferred. Since I needed to use `@` only twice for `condexp_L2` I did not change the definition to make it explicit.

Reply:
There is no issue if condexp_L2 is applied to a function (E and the measure are the ones from the function)